### PR TITLE
feat(ecmascript): implement `Date` constructor and prototype methods

### LINKED ISSUE #139: Date.prototype builtin functions

### PR DESCRIPTION
**Pull request by kermanx**
This PR implements most of the `Date` constructor and prototype methods. Closes #139.

Currently only UTC timezone is supported. Some date string parsers have not been implemented.

Many of the implementations are ported from https://github.com/boa-dev/boa/blob/main/core/engine/src/builtins/date/utils.rs and https://github.com/boa-dev/boa/blob/main/core/engine/src/builtins/date/mod.rs

btw, how can I update `matrix.json` without running all the test262 on my local machine?

### CODE REVIEW

**Comment by aapoalas** [comment]
Metrics / expectations can be updated with these instructions: https://github.com/trynova/nova/blob/main/CONTRIBUTING.md#tests-in-prs

Basically, this command:
```bash
cargo build --profile release && cargo run --bin test262 --profile release -- -u
```

**Review by aapoalas — CHANGES_REQUESTED** [review]
Holy moly, amazing work! Thank you and massive props for taking on this large body of work!

I had a few issues here and there, and nits to pick. The most problematic issue is regarding the GC safety of some of the calls; I unfortunately haven't quite yet finished a comprehensive guide to working with the GC but a basic one can be found [here](https://github.com/trynova/nova/blob/main/GARBAGE_COLLECTOR.md) and I'm writing a (hopefully) more comprehensive one in a blog format over [here](https://github.com/trynova/website/pull/42).

Beyond that, I wonder if we couldn't avoid using `f64` quite so much since the millisecond numbers are always rounded down to integers and are always kept in the safe integer range. I would assume that would simplify and perhaps speed up a lot of the code working around dates.

But all in all, great work, you really knocked it off the park right off the bat with this one!

File: nova_vm/src/ecmascript/builtins/numbers_and_dates/date_objects/date_prototype.rs [hunk 1]
```diff
@@ -883,3 +1378,973 @@ fn check_date_object<'a>(
...
+fn day_within_year(t: f64) -> u16 {
+    // 1. Return Day(t) - DayFromYear(YearFromTime(t)).
+    (day(t) - day_from_year(year_from_time(t).into())) as u16
+}
+
+/// ### [21.4.1.10 InLeapYear ( t )](https://tc39.es/ecma262/#sec-inleapyear)
+///
+/// The abstract operation InLeapYear takes argument t (a finite time value)
+/// and returns +0𝔽 or 1𝔽. It returns 1𝔽 if t is within a leap year and +0𝔽 otherwise.
+fn in_leap_year(t: f64) -> u16 {
```
[the remark this answers is not in the document]
Comment (kermanx) [reply nova_vm/src/ecmascript/builtins/numbers_and_dates/date_objects/date_prototype.rs:1483]: I think `u16` are better because the standard says it is a number and the return value is always used as an addend.

File: nova_vm/src/ecmascript/builtins/date/data.rs [hunk 2]
```diff
@@ -6,19 +6,18 @@ use crate::{
     ecmascript::types::OrdinaryObject,
     heap::{CompactionLists, HeapMarkAndSweep, WorkQueues},
 };
-use std::time::SystemTime;
 
 #[derive(Debug, Clone, Copy)]
 pub struct DateHeapData {
     pub(crate) object_index: Option<OrdinaryObject<'static>>,
-    pub(crate) date: Option<SystemTime>,
+    pub(crate) date: f64,
```
[the remark this answers is not in the document]
Comment (kermanx) [reply nova_vm/src/ecmascript/builtins/date/data.rs:13]: I borrowed this representation from Boa. I'm not sure why they use f64, maybe because the need to handle super large values? (but precision is still lost)

Should we change it to `i64`?

**Comment by kermanx** [comment]
> The most problematic issue is regarding the GC safety of some of the calls; I unfortunately haven't quite yet finished a comprehensive guide to working with the GC but a basic one can be found [here](https://github.com/trynova/nova/blob/main/GARBAGE_COLLECTOR.md?rgh-link-date=2025-03-09T13%3A12%3A12Z) and I'm writing a (hopefully) more comprehensive one in a blog format over [here](https://github.com/trynova/website/pull/42).

I read the blog posts and `GARBAGE_COLLECTOR.md` before PRing, but I didn't fully understand it. After reading the new draft of the blog post, I think I now have a better understanding of the idea.

File: nova_vm/src/ecmascript/builtins/date/data.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (aapoalas) [reply nova_vm/src/ecmascript/builtins/date/data.rs:13]: Hmm... V8 using an f64 is sort of a strong argument against using i64... But I do have a feeling that in our case using an `i64` is the correct choice: V8 only supports up to `i31` or `i32` as SMI (so on the stack), whereas we (at least currently) support the full `i53` range that JavaScript numbers are allowed to support. Date objects are guaranteed / required to only support a slightly smaller range of values:

> A Number can exactly represent all [integers](https://tc39.es/ecma262/#integer) from -9,007,199,254,740,992 to 9,007,199,254,740,992 ([21.1.2.8](https://tc39.es/ecma262/#sec-number.min_safe_integer) and [21.1.2.6](https://tc39.es/ecma262/#sec-number.max_safe_integer)). A time value supports a slightly smaller range of -8,640,000,000,000,000 to 8,640,000,000,000,000 milliseconds.

[Source](https://tc39.es/ecma262/#sec-time-values-and-time-range)

Those are thus all fully safe integers, so they're all valid i64's. And, importantly, because they're all safe integers then when we convert them into a `Value` in eg. `Date.prototype.toValue`, the conversion function will check if the `f64` is an integer within the safe integer range (always true for dates as seen above), and if so will convert it into an `i64` and then into a `SmallInteger` from that.

So, by storing dates as `f64` we're only adding to the work we need to do when converting them to `Value`, instead of making the work less like one would except from a usual JavaScript engine.

**Comment by kermanx** [comment]
> Metrics / expectations can be updated with these instructions: [`main`/CONTRIBUTING.md#tests-in-prs](https://github.com/trynova/nova/blob/main/CONTRIBUTING.md?rgh-link-date=2025-03-09T11%3A59%3A55Z#tests-in-prs)

It seems that `cargo run --bin test262 --profile release -- -u` runs for ~30 minutes on my machine and still does not finish. Is that normal? 🥹

**Comment by aapoalas** [comment]
> > Metrics / expectations can be updated with these instructions: [`main`/CONTRIBUTING.md#tests-in-prs](https://github.com/trynova/nova/blob/main/CONTRIBUTING.md?rgh-link-date=2025-03-09T11%3A59%3A55Z#tests-in-prs)
> 
> It seems that `cargo run --bin test262 --profile release -- -u` runs for ~30 minutes on my machine and still does not finish. Is that normal? 🥹

No, that does not sound normal at all. On my machine running the tests is maybe two minutes on release mode, perhaps 5-7 minutes on debug.

What's your machine like?

**Comment by aapoalas** [comment]
> > Metrics / expectations can be updated with these instructions: [`main`/CONTRIBUTING.md#tests-in-prs](https://github.com/trynova/nova/blob/main/CONTRIBUTING.md?rgh-link-date=2025-03-09T11%3A59%3A55Z#tests-in-prs)
> 
> It seems that `cargo run --bin test262 --profile release -- -u` runs for ~30 minutes on my machine and still does not finish. Is that normal? 🥹

Here are the exact commands I run locally:

```bash
cargo build && cargo build --profile release
cargo run -r --bin test262 -- -u
```

**Comment by kermanx** [comment]
> Here are the exact commands I run locally:
> 
> ```shell
> cargo build && cargo build --profile release
> cargo run -r --bin test262 -- -u
> ```

I just ran the exact same commands and it takes 40~50 minutes to finish. Some information of my machine:

```txt
OS: Windows 11
CPU: 13th Gen Intel(R) Core(TM) i7-1360P   2.20 GHz
RAM: 32.0 GB
```

I guess the difference between Linux and Windows may cause the problem.

**Comment by r58Playz** [comment]
I have the exact same CPU model and test262 runs in 16 seconds on Arch Linux:
```
________________________________________________________
Executed in   16.04 secs    fish           external
   usr time   92.10 secs    0.00 micros   92.10 secs
   sys time  105.97 secs  800.00 micros  105.97 secs
```

File: nova_vm/src/ecmascript/builtins/date/data.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kermanx) [reply nova_vm/src/ecmascript/builtins/date/data.rs]: In the latest commit, I changed the representation of date value into `struct DateValue(i64)`. which can be either (also described in code comments):

- Invalid, which is presented as `i64::MAX`
- An integer in the range of -8,640,000,000,000,000 to 8,640,000,000,000,000,
  which is represented as a non-max `i64`, and can also fit in `SmallInteger`

And in time calculation, `f64` is still used for now due to the following reasons:

- Many computations require down-passing NaN. If we use `i64` this can be quite hard and may be slower, and will introduce many differences between the implementation and the standard. 
- Values from arguments are `f64`.

To convert `DateValue` into `i64`/`f64`: `DateValue::get_i64()`/`DateValue::get_f64()`

To make a `DateValue`: `DateValue::now()`/[`time_clip`](https://tc39.es/ecma262/#sec-timeclip)

btw, I am not sure which file to put the definition of `DateValue` into. Currently it is at nova_vm\src\ecmascript\builtins\date\data.rs.

File: nova_vm/src/ecmascript/builtins/date/data.rs [hunk 3]
```diff
@@ -2,22 +2,103 @@
...
+/// This yields a supported time value range of exactly -100,000,000 days
+/// to 100,000,000 days relative to midnight at the beginning of 1 January 1970 UTC.
+///
+/// In that case, the time value can be either:
+///
+/// - Invalid, which is presented as `i64::MAX`
+/// - An integer in the range of -8,640,000,000,000,000 to 8,640,000,000,000,000,
+///   which is represented as a non-max `i64`, and can also fit in `SmallInteger`
+#[derive(Debug, Clone, Copy)]
+pub struct DateValue(i64);
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/date/data.rs:27]: **nitpick (blocking):** Let's not make types (or AOs) public if we can avoid it. Better to have too small an API surface to being with, than too large. (Note: Any abstract operations implemented should always by default be `pub(crate)` at most for the same reason.)

File: nova_vm/src/ecmascript/abstract_operations/type_conversion.rs [hunk 4]
```diff
@@ -447,6 +447,38 @@ impl IntegerOrInfinity {
...
+        } else if number == f64::INFINITY {
+            // 3. If number is +∞𝔽, return +∞.
+            Self::POS_INFINITY
+        } else if number == f64::NEG_INFINITY {
+            // 4. If number is -∞𝔽, return -∞.
+            Self::NEG_INFINITY
+        } else {
+            // 5. Let integer be floor(abs(ℝ(number))).
+            // 6. If number < +0𝔽, set integer to -integer.
+            let integer = number.abs().floor().copysign(number) as i64;
```
[the remark this answers is not in the document]
Comment (aapoalas) [reply nova_vm/src/ecmascript/abstract_operations/type_conversion.rs:476]: I'm still a bit worried about this. It's not a problem with the Date use-cases since we're probably always dealing with safe integers then, but if someone finds this method later and uses it with a plain f64 without thinking about it, it might cause an issue.

Let's add a few tests for this in this file, how about?

**Review by aapoalas — CHANGES_REQUESTED** [review]
Some nits to pick still, and let's add unit tests for the f64 -> IntegerOrInfinity to test that it doesn't overflow with big f64 inputs.

I'm... not totally thrilled with f64 still remaining in the calculation. It seems to me like a `DateValue` kind of wrapper around i64 (or just using `DateValue` directly, though it's a bit wrong admittedly since it's supposed to represent milliseconds) for the functions would be a reasonable choice that would avoid the pain of "how to deal with NaN i64s?" while still allowing computation to proceed mostly without pain on the i64 paths.

But, this fine as well. Let's just get the few minor edges filed down and this is good to go! Great stuff, absolutely stellar work! Thank you <3

File: nova_vm/src/ecmascript/abstract_operations/type_conversion.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kermanx) [reply nova_vm/src/ecmascript/abstract_operations/type_conversion.rs:476]: After some thought, I redesigned this version of `to_integer_or_infinity`. Now it accepts `f64` and returns `f64`. Previously it returned `IntegerOrInfinity` and then the result was always converted to `f64`.

I also changed `number.abs().floor().copysign(number)` into `number.trunc()`, the effect should be the same AFAIK.

**Review by aapoalas — DISMISSED** [review]
Sounds good to me!
